src/SFA.DAS.TrackProgress/Application/Commands/CreateProgressSnapshotCommand.cs

```diff
@@ -0,0 +1,95 @@
+﻿using MediatR;
+using Microsoft.EntityFrameworkCore;
+using NServiceBus;
+using SFA.DAS.TrackProgress.Database;
+using SFA.DAS.TrackProgress.Messages.Commands;
+using SFA.DAS.TrackProgress.Models;
+
+namespace SFA.DAS.TrackProgress.Application.Commands;
+
+public record CreateProgressSnapshotCommand(long CommitmentsApprenticeshipId) : IRequest;
+
+public class CreateProgressSnapshotCommandHandler : IRequestHandler<CreateProgressSnapshotCommand>
+{
+    private readonly TrackProgressContext _context;
+    private readonly IMessageSession _messageSession;
+
+    public CreateProgressSnapshotCommandHandler(TrackProgressContext context, IMessageSession messageSession)
+    {
+        _context = context;
+        _messageSession = messageSession;
+    }
+
+    public async Task<Unit> Handle(CreateProgressSnapshotCommand request, CancellationToken cancellationToken)
+    {
+        var snapshot = await BuildSnapshot(
+            request.CommitmentsApprenticeshipId, _context.Progress, cancellationToken);
+
+        await SaveSnapshot(snapshot, cancellationToken);
+
+        await CacheKsbNames(snapshot, cancellationToken);
+
+        return Unit.Value;
+    }
+
+    private static async Task<CourseSnapshot> BuildSnapshot(
+        long commitmentsApprenticeshipId,
+        IQueryable<Progress> progress,
+        CancellationToken cancellationToken)
+    {
+        var events = await progress
+            .Where(x => x.Approval.ApprenticeshipId == commitmentsApprenticeshipId)
+            .OrderBy(x => x.CreatedOn)
+            .ToListAsync(cancellationToken);
+
+        var allKsbSubmissions = events
+            .SelectMany(x => x.ProgressData.Ksbs)
+            .GroupBy(x => x.Id);
+
+        var latestKsbSubmissions = allKsbSubmissions
+            .Select(x => x.Last());
```

Comment:
Does this ensure the last track progress submission is the value used as part of the snap shot. Will this depend on the order in which SQL returns the records?

Reply:
Yes, because it's OrderedBy submission date on line 42